File: src/main/java/com/bajdcc/LALR1/grammar/type/ConvertToInt.java
package com.bajdcc.LALR1.grammar.type;

import com.bajdcc.util.lexer.token.Token;
import com.bajdcc.util.lexer.token.TokenType;

/**
 * 【类型转换】转换为整数类型
 *
 * @author bajdcc
 */
public class ConvertToInt implements ITokenConventer {

	@Override
	public Token convert(Token token) {
		TokenType type = token.getType();
		switch (token.getType()) {
			case BOOL:
			case CHARACTER:
			case STRING:
			case DECIMAL:
			case INTEGER:
				token.setObj(getIntValue(token));
				if (token.getType() == TokenType.ERROR) {
					token.setType(type);
				} else {
					token.setType(TokenType.INTEGER);
				}
				break;
			default:
				break;
		}
		return token;
	}

	/**
	 * 强制整数转换（值）
	 *
	 * @param token 操作数
	 * @return 转换结果
	 */
	private static long getIntValue(Token token) {
		switch (token.getType()) {
			case BOOL:
				boolean bool = (boolean) token.getObj();
				return bool ? 1L : 0L;
			case CHARACTER:
				char ch = (char) token.getObj();
				return (long) ch;
			case STRING:
				String str = (String) token.getObj();
				try {
					return Long.parseLong(str);
				} catch (NumberFormatException e) {
					token.setType(TokenType.ERROR);
				}
				break;
			case INTEGER:
				return (long) token.getObj();
			case DECIMAL:
				double decimal = (double) token.getObj();
				return (long) decimal;
			default:
				break;
		}
		return 0L;
	}
}

File: src/main/kotlin/com/bajdcc/LALR1/grammar/type/ConvertToInt.kt
package com.bajdcc.LALR1.grammar.type

import com.bajdcc.util.lexer.token.Token
import com.bajdcc.util.lexer.token.TokenType

/**
 * 【类型转换】转换为整数类型
 *
 * @author bajdcc
 */
class ConvertToInt : ITokenConventer {

    override fun convert(token: Token): Token {
        val type = token.type
        when (token.type) {
            TokenType.BOOL, TokenType.CHARACTER, TokenType.STRING, TokenType.DECIMAL, TokenType.INTEGER -> {
                token.obj = getIntValue(token)
                if (token.type === TokenType.ERROR) {
                    token.type = type
                } else {
                    token.type = TokenType.INTEGER
                }
            }
            else -> {
            }
        }
        return token
    }

    /**
     * 强制整数转换（值）
     *
     * @param token 操作数
     * @return 转换结果
     */
    private fun getIntValue(token: Token): Long {
        when (token.type) {
            TokenType.BOOL -> {
                val bool = token.obj as Boolean
                return if (bool) 1L else 0L
            }
            TokenType.CHARACTER -> {
                val ch = token.obj as Char
                return ch.toLong()
            }
            TokenType.STRING -> {
                val str = token.obj as String?
                try {
                    return java.lang.Long.parseLong(str!!)
                } catch (e: NumberFormatException) {
                    token.type = TokenType.ERROR
                }

            }
            TokenType.INTEGER -> return token.obj as Long
            TokenType.DECIMAL -> {
                val decimal = token.obj as Double
                return decimal.toLong()
            }
            else -> {
            }
        }
        return 0L
    }
}
